transformer_engine/pytorch/module.py

```diff
@@ -300,22 +303,25 @@ def pre_forward(self, inp: torch.Tensor, num_gemms: int = 1) -> None:
         self.fp8_init(num_gemms=num_gemms)
         self.set_fp8_weights()
 
+        # Previous iteration was grad_enabled
         if self.fp8_meta.get("update_amax_and_scale_fwd", False):
-            # Previous iteration was grad_enabled
-            copy_amax_from_global_buffer(self.fp8_meta, forward=True)
-            amax_and_scale_update(self.fp8_meta, True)
-            set_amax_buffer_key_deletion(self.fp8_meta, forward=True)
-
-        if self.fp8 and torch.is_grad_enabled() and self.training:
-            self.fp8_meta["first_module"] = is_first_fp8_module()
-
-            if self.fp8_meta["first_module"]:
-                self.fp8_meta["autocast_id_fwd"] = new_fp8_context_id()
-                set_fp8_context_id(self.fp8_meta["autocast_id_fwd"])
+            if self.fp8_meta["recipe"].reduce_amax:
+                copy_amax_from_global_buffer(self.fp8_meta, forward=True)
+                amax_and_scale_update(self.fp8_meta, True)
+                set_amax_buffer_key_deletion(self.fp8_meta, forward=True)
             else:
-                self.fp8_meta["autocast_id_fwd"] = get_fp8_context_id()
+                amax_and_scale_update(self.fp8_meta, True)
 
-            add_amax_to_global_buffer(self.fp8_meta, forward=True)
+        if self.fp8 and torch.is_grad_enabled() and self.training:
+            # Setup for amax reduction
+            if self.fp8_meta["recipe"].reduce_amax:
+                self.fp8_meta["first_module"] = is_first_fp8_module()
```

Comment:
Why is this part skipped when not reducing amax? You still need to update the scaling factor in that case at the end of fwd and backward so I think the data on whether it is the first fp8 module should still be updated?

Reply:
Scaling factor and amax updates are local per module and only the reductions are aggregated. So information about first module isn't needed for this case.

Reply:
But how can you update the scaling factors if you don't have amax yet? If the amax reduction is aggregated then the scaling factors update also has to be, right?

Reply:
The aggregation of the amaxes for reduction is a separate step from updating the amax and scale factors. The latter is done at the beginning of the following iteration and is localized per module (not aggregated).